File: Исходники/Глава 2. Часть 1/Жадные алгоритмы/8. Алгоритм Краскала/C++/Продуктовый/MST/UnionFind.cpp
#include "UnionFind.h"

UnionFind::UnionFind(int vertex) {
	rank.assign(vertex, 0);
	p.assign(vertex, 0);
	for (int i = 0; i < vertex; i++) p[i] = i;
}

int UnionFind::find_set(int v) {
	if (p[v] == v) { return v; }
	else { return find_set(p[v]); }
}

bool UnionFind::same_set(int v1, int v2) {
	return find_set(v1) == find_set(v2);
}

void UnionFind::union_set(int v1, int v2) {
	if (!same_set(v1, v2)) {
		int x = find_set(v1);
		int y = find_set(v2);
		if (rank[x] > rank[y]) { p[y] = x; }
		else {
			p[x] = y;
			if (rank[x] == rank[y]) rank[y]++;
		}
	}
}

File: Исходники/Глава 2. Часть 1/Жадные алгоритмы/8. Алгоритм Краскала/Python/MST/UnionFind.py
class UnionFind:
    def __init__(self, n):
        self.p = [i for i in range(n)]
        self.rank = [0] * n

    def find_set(self, v):
        if self.p[v] == v:
            return v
        return self.find_set(self.p[v])

    def same_set(self, v1, v2):
        return self.find_set(self.p[v1]) == self.find_set(self.p[v2])

    def union_set(self, v1, v2):
        x = self.find_set(v1)
        y = self.find_set(v2)
        if self.rank[x] > self.rank[y]:
            self.p[y] = x
        else:
            self.p[x] = y
            if self.rank[x] == self.rank[y]:
                self.rank[y] += 1
